src/dowel/csv_output.py

```diff
@@ -30,26 +32,41 @@ def record(self, data, prefix=''):
         if isinstance(data, TabularInput):
             to_csv = data.as_primitive_dict
 
-            if not to_csv.keys() and not self._writer:
+            # if not to_csv.keys() and not self._writer:
+            #     return
+
+            # if not self._writer:
+            #     self._fieldnames = set(to_csv.keys())
+            #     self._writer = csv.DictWriter(
+            #         self._log_file,
+            #         fieldnames=self._fieldnames,
+            #         extrasaction='ignore')
+            #     self._writer.writeheader()
+
+            # if to_csv.keys() != self._fieldnames:
+            #     self._warn('Inconsistent TabularInput keys detected. '
+            #                'CsvOutput keys: {}. '
+            #                'TabularInput keys: {}. '
+            #                'Did you change key sets after your first '
+            #                'logger.log(TabularInput)?'.format(
+            #                    set(self._fieldnames), set(to_csv.keys())))
+
+            # self._writer.writerow(to_csv)
+
+            if not to_csv.keys():
                 return
 
-            if not self._writer:
-                self._fieldnames = set(to_csv.keys())
-                self._writer = csv.DictWriter(
-                    self._log_file,
-                    fieldnames=self._fieldnames,
-                    extrasaction='ignore')
-                self._writer.writeheader()
-
-            if to_csv.keys() != self._fieldnames:
-                self._warn('Inconsistent TabularInput keys detected. '
-                           'CsvOutput keys: {}. '
-                           'TabularInput keys: {}. '
-                           'Did you change key sets after your first '
-                           'logger.log(TabularInput)?'.format(
-                               set(self._fieldnames), set(to_csv.keys())))
-
-            self._writer.writerow(to_csv)
+            # Add new data, Pandas will create new columns automatically
+            self._df = self._df.append(to_csv, ignore_index=True)
+            temp = pd.DataFrame().append(to_csv, ignore_index=True)
+
+            if self._fieldnames != set(to_csv.keys()):  # need to rewrite all
+                self._log_file.seek(0, 0)  # reset the pointer
+                self._df.to_csv(self._log_file, index=False)  # write data
+                self._fieldnames = set(to_csv.keys())  # renew cols
+            else:  # just append the new row
+                temp.to_csv(
+                    self._log_file, mode='a', header=False, index=False)
```

Comment:
How do we know that the order of columns is consistent between calls here? The `TabularInput` might have keys in a different order.

Reply:
You are right. For the keys which are the same as a set, they might be different when being written to CSV files due to the random order. This needs to be fixed.

Reply:
My new test code:
```
import dowel
from dowel import logger, tabular

logger.add_output(dowel.StdOutput())
logger.add_output(dowel.CsvOutput('out.csv'))

tabular.record('itr', 1)
tabular.record('loss', 0.5)
logger.log(tabular)
logger.dump_all()

tabular.record('loss', 0.3)
tabular.record('itr', 2)
logger.log(tabular)
logger.dump_all()

tabular.record('itr', 3)
tabular.record('loss', 0.1)
tabular.record('new', 0.77)
logger.log(tabular)
logger.dump_all()

tabular.record('new', 0.88)
logger.log(tabular)
logger.dump_all()

logger.remove_all()
```

And the result:
<img width="219" alt="image" src="https://user-images.githubusercontent.com/28439048/82506570-06ad4c00-9ab5-11ea-9cab-51c5b7d48db4.png">

It seems working well.